File: clients/Java/src/main/java/MyStrategy.java
import model.*;

public class MyStrategy {
    public Action getAction(PlayerView playerView, DebugInterface debugInterface) {
        Action result = new Action(new java.util.HashMap<>());
        int myId = playerView.getMyId();
        for (Entity entity : playerView.getEntities()) {
            if (entity.getPlayerId() == null || entity.getPlayerId() != myId) {
                continue;
            }
            EntityProperties properties = playerView.getEntityProperties().get(entity.getEntityType());

            MoveAction moveAction = null;
            BuildAction buildAction = null;
            if (properties.isCanMove()) {
                moveAction = new MoveAction(
                    new Vec2Int(playerView.getMapSize() - 1, playerView.getMapSize() - 1),
                    true,
                    true);
            } else if (properties.getBuild() != null) {
                EntityType entityType = properties.getBuild().getOptions()[0];
                int currentUnits = 0;
                for (Entity otherEntity : playerView.getEntities()) {
                    if (otherEntity.getPlayerId() != null && otherEntity.getPlayerId() == myId
                        && otherEntity.getEntityType() == entityType) {
                        currentUnits++;
                    }
                }
                if ((currentUnits + 1) * playerView.getEntityProperties().get(entityType).getPopulationUse() <= properties.getPopulationProvide()) {
                    buildAction = new BuildAction(
                        entityType,
                        new Vec2Int(
                            entity.getPosition().getX() + properties.getSize(),
                            entity.getPosition().getY() + properties.getSize() - 1
                        )
                    );
                }
            }
            EntityType[] validAutoAttackTargets;
            if (entity.getEntityType() == EntityType.BUILDER_UNIT) {
                validAutoAttackTargets = new EntityType[] { EntityType.RESOURCE };
            } else {
                validAutoAttackTargets = new EntityType[0];
            }
            result.getEntityActions().put(entity.getId(), new EntityAction(
                moveAction,
                buildAction,
                new AttackAction(
                    null, new AutoAttack(properties.getSightRange(), validAutoAttackTargets)
                ),
                null
            ));
        }
        return result;
    }
    public void debugUpdate(PlayerView playerView, DebugInterface debugInterface) {
        debugInterface.send(new DebugCommand.Clear());
        debugInterface.getState();
    }
}

File: clients/Rust/src/my_strategy.rs
use super::DebugInterface;
use model::*;

pub struct MyStrategy {}

impl MyStrategy {
    pub fn new() -> Self {
        Self {}
    }
    pub fn get_action(
        &mut self,
        player_view: &PlayerView,
        debug_interface: Option<&mut DebugInterface>,
    ) -> Action {
        let mut result = Action {
            entity_actions: std::collections::HashMap::new(),
        };
        let my_id = player_view.my_id;
        for entity in player_view
            .entities
            .iter()
            .filter(|entity| entity.player_id == Some(my_id))
        {
            let properties = &player_view.entity_properties[&entity.entity_type];

            let mut move_action = None;
            let mut build_action = None;
            if properties.can_move {
                move_action = Some(MoveAction {
                    target: Vec2I32 {
                        x: player_view.map_size - 1,
                        y: player_view.map_size - 1,
                    },
                    find_closest_position: true,
                    break_through: true,
                });
            } else if let Some(build_properties) = &properties.build {
                let entity_type = &build_properties.options[0];
                let current_units = player_view
                    .entities
                    .iter()
                    .filter(|entity| {
                        entity.player_id == Some(my_id) && entity.entity_type == *entity_type
                    })
                    .count() as i32;
                if (current_units + 1) * player_view.entity_properties[&entity_type].population_use
                    <= properties.population_provide
                {
                    build_action = Some(BuildAction {
                        entity_type: entity_type.clone(),
                        position: Vec2I32 {
                            x: entity.position.x + properties.size,
                            y: entity.position.y + properties.size - 1,
                        },
                    });
                }
            }
            result.entity_actions.insert(
                entity.id,
                EntityAction {
                    move_action,
                    build_action,
                    attack_action: Some(AttackAction {
                        target: None,
                        auto_attack: Some(AutoAttack {
                            pathfind_range: properties.sight_range,
                            valid_targets: if entity.entity_type == EntityType::BuilderUnit {
                                vec![EntityType::Resource]
                            } else {
                                vec![]
                            },
                        }),
                    }),
                    repair_action: None,
                },
            );
        }
        result
    }
    pub fn debug_update(&mut self, player_view: &PlayerView, debug_interface: &mut DebugInterface) {
        debug_interface.send(DebugCommand::Clear {});
        debug_interface.get_state();
    }
}
